test/integration/oauth_cert_fallback_test.go

```diff
@@ -0,0 +1,208 @@
+3Yt5D2gTLJ2MUpMxWv+cFz/UJZ0TBKtoKLjtLGRCGwFcUOz6rJzM7QGChfQxMzD7
+gydO+4blIu/i4strKbkR5jcRA1WgwS/1yW25F/hE0QglsWyXJJELf4ZCNpr+9zQb
+eUArEOSZyp5WmmRcn0v/YZc=
+-----END CERTIFICATE-----`)
+
+		invalidToken = "invalid"
+		noToken      = ""
+
+		invalidCert = restclient.TLSClientConfig{
+			CertData: invalidCertData,
+			KeyData:  invalidKeyData,
+		}
+		noCert = restclient.TLSClientConfig{}
+
+		tokenUser = "user"
+		certUser  = "system:admin"
+
+		unauthorizedError = "the server has asked for the client to provide credentials (get users ~)"
+		anonymousError    = `User "system:anonymous" cannot get users at the cluster scope`
+	)
+
+	// TODO see comment below
+	_ = invalidCert
+
+	testutil.RequireEtcd(t)
+	// Build master config
+	masterOptions, err := testserver.DefaultMasterOptions()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Start server
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMaster(masterOptions)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	adminConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	validCert := adminConfig.TLSClientConfig
+
+	validToken, err := tokencmd.RequestToken(adminConfig, nil, tokenUser, "pass")
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	if len(validToken) == 0 {
+		t.Fatalf("Expected valid token, got none")
+	}
+
+	for _, test := range []struct {
+		token         string
+		cert          restclient.TLSClientConfig
+		expectedUser  string
+		errorExpected bool
+		errorString   string
+		insecure      bool
+	}{
+		{
+			token:        validToken,
+			cert:         validCert,
+			expectedUser: tokenUser,
+		},
+		// TODO find a way to test invalidCert
+		// Tests with invalidCert don't work because go does not send the client cert
+		// because it knows the server does not want certs that are not signed it's CA
+		// {
+		// 	token:        validToken,
+		// 	cert:         invalidCert,
+		// 	expectedUser: tokenUser,
+		// 	insecure:     true,
+		// },
+		{
+			token:        validToken,
+			cert:         noCert,
+			expectedUser: tokenUser,
+			insecure:     true,
+		},
+		{
+			token:         invalidToken,
+			cert:          validCert,
+			errorExpected: true,
+			errorString:   unauthorizedError,
+		},
+		// {
+		// 	token:         invalidToken,
+		// 	cert:          invalidCert,
+		// 	errorExpected: true,
+		// 	errorString:   unauthorizedError,
+		// 	insecure:      true,
+		// },
+		{
+			token:         invalidToken,
+			cert:          noCert,
+			errorExpected: true,
+			errorString:   unauthorizedError,
+			insecure:      true,
+		},
+		{
+			token:        noToken,
+			cert:         validCert,
+			expectedUser: certUser,
+		},
+		// {
+		// 	token:         noToken,
+		// 	cert:          invalidCert,
+		// 	errorExpected: true,
+		// 	errorString:   anonymousError, // TODO should this be unauthorizedError?
```

Comment:
Should be unauthorized error once we figure out how to actually test the scenario